File: src/main/java/com/github/kory33/signvote/io/PluginDataAutoSaver.java
package com.github.kory33.signvote.io;

import com.github.kory33.signvote.core.SignVote;
import org.bukkit.Server;
import org.bukkit.scheduler.BukkitScheduler;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

/**
 * A class which is capable of scheduling auto-save of the plugin data with a specified time interval.
 * @author Kory
 */
public class PluginDataAutoSaver {
    private int nextAutoSaveTaskId;

    private final long autosaveIntervalTicks;
    private final boolean shouldLog;

    private final SignVote plugin;
    private final Server server;
    private final BukkitScheduler scheduler;
    private final ExecutorService saveTaskExecutor;
    
    private void scheduleNextAutoSaveTask() {
        CompletableFuture.runAsync(PluginDataAutoSaver.this.plugin::saveSessionData, PluginDataAutoSaver.this.saveTaskExecutor);
        if (this.shouldLog) {
            PluginDataAutoSaver.this.plugin.getLogger().info("Session data is being saved asynchronously...");
        }
        
        this.nextAutoSaveTaskId = scheduler.scheduleSyncDelayedTask(this.plugin,
                        PluginDataAutoSaver.this::scheduleNextAutoSaveTask, this.autosaveIntervalTicks);
    }

    /**
     * Stop and cancel the autosave task.
     */
    public void stopAutoSaveTask() {
        server.getScheduler().cancelTask(this.nextAutoSaveTaskId);
    }
    
    public PluginDataAutoSaver(SignVote plugin, int autosaveInterval, boolean shouldLog) {
        this.plugin = plugin;
        this.server = plugin.getServer();
        this.scheduler = server.getScheduler();

        this.autosaveIntervalTicks = autosaveInterval;
        this.saveTaskExecutor = Executors.newFixedThreadPool(1);
        this.shouldLog = shouldLog;
        
        this.scheduleNextAutoSaveTask();
    }
}

File: src/main/kotlin/com/github/kory33/signvote/io/PluginDataAutoSaver.kt
package com.github.kory33.signvote.io

import com.github.kory33.signvote.core.SignVote
import org.bukkit.Server
import org.bukkit.scheduler.BukkitScheduler

import java.util.concurrent.CompletableFuture
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

/**
 * A class which is capable of scheduling auto-save of the plugin data with a specified time interval.
 * @author Kory
 */
class PluginDataAutoSaver(private val plugin: SignVote, autosaveInterval: Int, private val shouldLog: Boolean) {
    private var nextAutoSaveTaskId: Int = 0

    private val autosaveIntervalTicks: Long
    private val server: Server = plugin.server
    private val scheduler: BukkitScheduler
    private val saveTaskExecutor: ExecutorService

    private fun scheduleNextAutoSaveTask() {
        CompletableFuture.runAsync(Runnable { this.plugin.saveSessionData() }, this.saveTaskExecutor)
        if (this.shouldLog) {
            this.plugin.logger.info("Session data is being saved asynchronously...")
        }

        this.nextAutoSaveTaskId = scheduler.scheduleSyncDelayedTask(this.plugin,
                { this.scheduleNextAutoSaveTask() }, this.autosaveIntervalTicks)
    }

    /**
     * Stop and cancel the autosave task.
     */
    fun stopAutoSaveTask() {
        server.scheduler.cancelTask(this.nextAutoSaveTaskId)
    }

    init {
        this.scheduler = server.scheduler

        this.autosaveIntervalTicks = autosaveInterval.toLong()
        this.saveTaskExecutor = Executors.newFixedThreadPool(1)

        this.scheduleNextAutoSaveTask()
    }
}
